src/lib/kernel_npem.c

```diff
@@ -0,0 +1,205 @@
+
+/* System headers */
+#include <stdio.h>
+#include <string.h>
+#include <pci/pci.h>
+#include <time.h>
+#include <stdbool.h>
+#include <sys/stat.h>
+
+/* Local headers */
+#include "libled_private.h"
+#include "cntrl.h"
+#include "npem.h"
+#include "utils.h"
+
+/* NPEM OK Capable/Control */
+#define PCI_NPEM_OK_CAP		0x004
+/* NPEM Locate Capable/Control */
+#define PCI_NPEM_LOCATE_CAP	0x008
+/* NPEM Fail Capable/Control */
+#define PCI_NPEM_FAIL_CAP	0x010
+/* NPEM Rebuild Capable/Control */
+#define PCI_NPEM_REBUILD_CAP	0x020
+/* NPEM Predicted Failure Analysis Capable/Control */
+#define PCI_NPEM_PFA_CAP	0x040
+/* NPEM Hot Spare Capable/Control */
+#define PCI_NPEM_HOT_SPARE_CAP	0x080
+/* NPEM in a Critical Array Capable/Control */
+#define PCI_NPEM_CRA_CAP	0x100
+/* NPEM in a Failed Array Capable/Control */
+#define PCI_NPEM_FA_CAP		0x200
+
+static const struct ibpi2value ibpi_to_npem_capability[] = {
+	{LED_IBPI_PATTERN_NORMAL, PCI_NPEM_OK_CAP},
+	{LED_IBPI_PATTERN_ONESHOT_NORMAL, PCI_NPEM_OK_CAP},
+	{LED_IBPI_PATTERN_DEGRADED, PCI_NPEM_CRA_CAP},
+	{LED_IBPI_PATTERN_HOTSPARE, PCI_NPEM_HOT_SPARE_CAP},
+	{LED_IBPI_PATTERN_REBUILD, PCI_NPEM_REBUILD_CAP},
+	{LED_IBPI_PATTERN_FAILED_ARRAY, PCI_NPEM_FA_CAP},
+	{LED_IBPI_PATTERN_PFA, PCI_NPEM_PFA_CAP},
+	{LED_IBPI_PATTERN_FAILED_DRIVE, PCI_NPEM_FAIL_CAP},
+	{LED_IBPI_PATTERN_LOCATE, PCI_NPEM_LOCATE_CAP},
+	{LED_IBPI_PATTERN_LOCATE_OFF, PCI_NPEM_OK_CAP},
+	{LED_IBPI_PATTERN_UNKNOWN, 0}
+};
+
+struct kernel_npem_led {
+	int	bitmask;
+	char	*sysfs_led_name;
+};
+
+const struct kernel_npem_led kernel_npem_leds[] = {
+	{PCI_NPEM_OK_CAP, "enclosure:ok"},
+	{PCI_NPEM_LOCATE_CAP, "enclosure:locate"},
+	{PCI_NPEM_FAIL_CAP, "enclosure:fail"},
+	{PCI_NPEM_REBUILD_CAP, "enclosure:rebuild"},
+	{PCI_NPEM_PFA_CAP, "enclosure:pfa"},
+	{PCI_NPEM_HOT_SPARE_CAP, "enclosure:hotspare"},
+	{PCI_NPEM_CRA_CAP, "enclosure:ica"},
+	{PCI_NPEM_FA_CAP, "enclosure:ifa"},
+};
+
+char *kernel_npem_get_path(const char *cntrl_path)
+{
+	return strdup(cntrl_path);
+}
+
+#define make_led_path(sysfs_led_path, sysfs_path, sysfs_led_name) \
+	snprintf(sysfs_led_path, sizeof(sysfs_led_path), "%s/leds/%s:%s/brightness", \
+		 sysfs_path, basename(sysfs_path), sysfs_led_name)
+
+static u32 read_kernel_npem_register(const char *sysfs_path)
+{
+	char led_path[PATH_MAX];
+	int i;
+	u32 reg = 0;
+
+	for (i = 0; i < ARRAY_SIZE(kernel_npem_leds); i++) {
+		make_led_path(led_path, sysfs_path, kernel_npem_leds[i].sysfs_led_name);
+		reg |= get_int("/", 0, led_path) ? kernel_npem_leds[i].bitmask : 0;
+	}
+	return reg;
+}
+
+static int write_kernel_npem_register(const char *sysfs_path, u32 val)
+{
+	char led_path[PATH_MAX], val_text[4];
+	int i;
+	struct stat sb;
+
+	for (i = 0; i < ARRAY_SIZE(kernel_npem_leds); i++) {
+		make_led_path(led_path, sysfs_path, kernel_npem_leds[i].sysfs_led_name);
+		snprintf(val_text, sizeof(val_text), val & kernel_npem_leds[i].bitmask ? "1" : "0");
+		if (!stat(led_path, &sb))
+			buf_write(led_path, val_text);
```

Comment:
not checking return code of `buf_write(led_path, val_text);` , but I suppose it failing is remote here.